### ISSUE TITLE
Incorrect output for 32 bit windows

### ISSUE THREAD

**Issue by richfitz** [richfitz/dde#14]
To see the problem on appveyor, run:

```r
deriv <- function(t, y, p) {
  1
}
ans1 <- dde::dopri(0, c(0, 1), deriv, 0, return_statistics = TRUE,
                   verbose = TRUE)
print(ans1)
print(drop(dde::dopri(0, c(0, 1), deriv, 0, return_minimal = TRUE)))
```

And then look at the output - the end number is never set.

Without the `memset` line in `r_dopri.c` this returns uninitialised memory so for some reason we are just never setting this time, but we do  actually run things:

```
t: 0.000000, h: 1.000000e-04, accept: yes
t: 0.000100, h: 1.000000e-03, accept: yes
t: 0.001100, h: 1.000000e-02, accept: yes
t: 0.011100, h: 1.000000e-01, accept: yes
t: 0.111100, h: 8.889000e-01, accept: yes
     [,1] [,2]
[1,]    0    0
[2,]    1    0
attr(,"statistics")
  n_eval   n_step n_accept n_reject 
      32        5        5        0 
attr(,"step_size")
[1] 0.8889
```

These numbers all look exactly the same as the 64 bit version.

To expose this bug, remove the `skip_on_appveyor` call from `integrate function with no absolute error` in `test-ode.R`

**Comment by weshinsley** [comment]
https://github.com/mrc-ide/dde/pull/16

**Comment by weshinsley** [comment]
This issue turns out to be a optimisation-related bug observed in gcc 4.9.3, 32-bit on Windows, which comes with Rtools 3.5. It does not occur with compiler flag -O0 , but does occur for -O1, -O2 or -O3.

The while loop around line 417 of src/dopri.c the original code gets optimised in a way that causes fewer iterations to occur for the failing example. Evaluating the elements of the while loop one by one can (at times) reveal them all to be true, showing that the while loop should never have been exited from.

The PR addresses this in a roundabout way with alternative logic that uses a variable some distance away to decide whether the loop should iterate. This seems enough to prevent gcc from introducing the bug.

https://github.com/mrc-ide/dde/pull/19

**Comment by richfitz** [comment]
Fixed in #19
